# Question 1
# Which type of complaint should the Department of Housing Preservation and Development of New York City focus on first?

# Data Download

In [2]:
#Bucket and Credentials

#import types
import pandas as pd
from botocore.client import Config
import ibm_boto3
#def __iter__(self): return 0

client_cred = ibm_boto3.client(service_name='s3',
ibm_api_key_id='rA-GmnuRkJ81A9EsDqha63ZmdFzQjxwCEgASTF2F-OJr',
ibm_auth_endpoint='https://iam.cloud.ibm.com/oidc/token',
config=Config(signature_version='oauth'),
endpoint_url='https://s3-api.us-geo.objectstorage.service.networklayer.com')

bucket = 'ibmdatascienceandmachinelearningc-donotdelete-pr-bkscl1den6u69w'

In [3]:
# Download NYC 311 Calls Dataset

client_cred.download_file(Bucket=bucket, Key='df_raw_cos.pkl', Filename='./df_raw_local.pkl')
df = pd.read_pickle('./df_raw_local.pkl')
df.head()

created_date  unique_key        complaint_type  incident_zip  \
0  2020-07-02T00:53:46.000    46724396  UNSANITARY CONDITION       10453.0   
1  2020-07-02T01:30:47.000    46723307              ELECTRIC       11377.0   
2  2020-07-02T01:51:58.000    46720951              ELECTRIC       10460.0   
3  2020-07-02T01:30:28.000    46721660  UNSANITARY CONDITION       10030.0   
4  2020-07-02T01:48:56.000    46715042            WATER LEAK       10456.0   

                    incident_address                   street_name  \
0               1735 DAVIDSON AVENUE               DAVIDSON AVENUE   
1                    51-28 72 STREET                     72 STREET   
2                1967 MARMION AVENUE                MARMION AVENUE   
3  2500 FREDERICK DOUGLASS BOULEVARD  FREDERICK DOUGLASS BOULEVARD   
4               1100 FRANKLIN AVENUE               FRANKLIN AVENUE   

  address_type      city                             resolution_description  \
0      ADDRESS     BRONX  The following complaint conditions are still o...   
1      ADDRESS  Woodside  The following complaint conditions are still o...   
2      ADDRESS     BRONX  The following complaint conditions are still o...   
3      ADDRESS  NEW YORK  The following complaint conditions are still o...   
4      ADDRESS     BRONX  The following complaint conditions are still o...   

     borough   latitude  longitude closed_date         location_type status  
0      BRONX  40.847996 -73.913303         NaN  RESIDENTIAL BUILDING   Open  
1     QUEENS  40.734983 -73.892334         NaN  RESIDENTIAL BUILDING   Open  
2      BRONX  40.843515 -73.886672         NaN  RESIDENTIAL BUILDING   Open  
3  MANHATTAN  40.815557 -73.947222         NaN  RESIDENTIAL BUILDING   Open  
4      BRONX  40.827819 -73.906298         NaN  RESIDENTIAL BUILDING   Open

In [4]:
# Download NYC PLUTO file Brooklyn 

client_cred.download_file(Bucket=bucket, Key='df_data_4_raw_cos.pkl', Filename='./df1.pkl')
df_BR = pd.read_pickle('./df1.pkl')
df_BR.head()

Borough  Block   Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      BK      1     1  302    21.0     NaN        13.0     33.0  11201.0   
1      BK      1    50  302    21.0  2000.0        13.0     33.0  11201.0   
2      BK      1  7501  302    21.0  2000.0        13.0     33.0  11201.0   
3      BK      3     1  302    21.0  3002.0        13.0     33.0  11201.0   
4      BK      3     5  302    21.0     NaN        13.0     33.0  11201.0   

  FireComp  ...  ZMCode  Sanborn   TaxMap  EDesigNum        APPBBL  \
0     L118  ...     NaN  302 007  30101.0        NaN  3.000010e+09   
1     L118  ...     NaN  302 007  30101.0      E-231  0.000000e+00   
2     L118  ...     NaN  302 007  30101.0        NaN  3.000010e+09   
3     L118  ...     NaN  302 007  30101.0        NaN  0.000000e+00   
4     L118  ...     NaN  302 007  30101.0        NaN  0.000000e+00   

      APPDate PLUTOMapID FIRM07_FLAG PFIRM15_FLAG Version  
0  11/26/2013          1         1.0          1.0    18V1  
1         NaN          1         1.0          1.0    18V1  
2  03/04/2016          1         1.0          1.0    18V1  
3         NaN          1         1.0          1.0    18V1  
4         NaN          4         1.0          1.0    18V1  

[5 rows x 87 columns]

In [5]:
# Download NYC PLUTO file Bronx

client_cred.download_file(Bucket=bucket, Key='df_data_5_raw_cos.pkl', Filename='./df1.pkl')
df_BX = pd.read_pickle('./df1.pkl')
df_BX.head()

Borough  Block  Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      BX   2260    1  201    19.0  1022.0         7.0      8.0  10454.0   
1      BX   2260    4  201    19.0  1022.0         7.0      8.0  10454.0   
2      BX   2260   10  201    19.0  1022.0         7.0      8.0  10454.0   
3      BX   2260   17  201    19.0  1022.0         7.0      8.0  10454.0   
4      BX   2260   18  201    19.0  1022.0         7.0      8.0  10454.0   

  FireComp  ...  ZMCode  Sanborn   TaxMap  EDesigNum  APPBBL APPDate  \
0     L029  ...     NaN  209S016  20901.0      E-143     0.0     NaN   
1     L029  ...     NaN  209S016  20901.0      E-143     0.0     NaN   
2     L029  ...     NaN  209S016  20901.0      E-143     0.0     NaN   
3     L029  ...     NaN  209S016  20901.0      E-143     0.0     NaN   
4     L029  ...     NaN  209S016  20901.0      E-143     0.0     NaN   

  PLUTOMapID FIRM07_FLAG PFIRM15_FLAG Version  
0          1         NaN          NaN    18V1  
1          1         NaN          NaN    18V1  
2          1         NaN          NaN    18V1  
3          1         NaN          NaN    18V1  
4          1         NaN          NaN    18V1  

[5 rows x 87 columns]

In [6]:
# Download NYC PLUTO file Manhatan

client_cred.download_file(Bucket=bucket, Key='df_data_6_raw_cos.pkl', Filename='./df1.pkl')
df_MN = pd.read_pickle('./df1.pkl')
df_MN.head()

Borough  Block  Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      MN      1   10  101     5.0  1011.0         2.0      1.0  10004.0   
1      MN      1  101  101     1.0  1001.0         2.0      1.0  10004.0   
2      MN      1  201  101     1.0  1000.0         2.0      1.0  10004.0   
3      MN      1  301  101     NaN     NaN         2.0      1.0  10004.0   
4      MN      1  401  101     NaN     NaN         2.0      1.0  10004.0   

  FireComp  ...  ZMCode   Sanborn   TaxMap  EDesigNum  APPBBL APPDate  \
0     E007  ...       Y   199 999  10101.0        NaN     0.0     NaN   
1     E007  ...       Y   199 999  10101.0        NaN     0.0     NaN   
2     E007  ...     NaN   199 999  10101.0        NaN     0.0     NaN   
3     E007  ...     NaN   199 999  10101.0        NaN     0.0     NaN   
4     E007  ...     NaN  1 99 999  10101.0        NaN     0.0     NaN   

  PLUTOMapID FIRM07_Flag PFIRM15_Flag Version  
0          1         1.0          1.0    18V1  
1          1         NaN          1.0    18V1  
2          1         NaN          1.0    18V1  
3          4         1.0          1.0    18V1  
4          4         1.0          1.0    18V1  

[5 rows x 87 columns]

In [7]:
# Download NYC PLUTO file Queens

client_cred.download_file(Bucket=bucket, Key='df_data_7_raw_cos.pkl', Filename='./df1.pkl')
df_QN = pd.read_pickle('./df1.pkl')
df_QN.head()

Borough  Block  Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      QN      6    1  402     1.0     NaN        30.0     26.0  11101.0   
1      QN      6    3  402     1.0  1015.0        30.0     26.0  11101.0   
2      QN      6    8  402     1.0  1011.0        30.0     26.0  11101.0   
3      QN      6   20  402     1.0     NaN        30.0     26.0  11101.0   
4      QN      6   30  402     1.0     NaN        30.0     26.0  11101.0   

  FireComp  ...  ZMCode  Sanborn   TaxMap  EDesigNum        APPBBL  \
0     L115  ...       Y  401 011  40101.0        NaN  4.000060e+09   
1     L115  ...     NaN  401 011  40101.0        NaN  0.000000e+00   
2     L115  ...     NaN  401 011  40101.0        NaN  4.000060e+09   
3     L115  ...     NaN  401 011  40101.0        NaN  4.000060e+09   
4     L115  ...     NaN  401 011  40101.0        NaN  4.000060e+09   

      APPDate PLUTOMapID FIRM07_FLAG PFIRM15_FLAG Version  
0  09/20/2013          1         1.0          1.0    18V1  
1         NaN          1         1.0          1.0    18V1  
2  08/07/2013          1         1.0          1.0    18V1  
3  09/20/2013          1         1.0          1.0    18V1  
4  09/08/2017          1         1.0          1.0    18V1  

[5 rows x 87 columns]

In [8]:
# Download NYC PLUTO file Staten Island

client_cred.download_file(Bucket=bucket, Key='df_data_8_raw_cos.pkl', Filename='./df1.pkl')
df_SI = pd.read_pickle('./df1.pkl')
df_SI.head()

Borough  Block  Lot   CD  CT2010  CB2010  SchoolDist  Council  ZipCode  \
0      SI      1   10  501     3.0  2000.0        31.0     49.0  10301.0   
1      SI      1   17  501     3.0     NaN        31.0     49.0  10301.0   
2      SI      1   18  501     3.0     NaN        31.0     49.0  10301.0   
3      SI      1   19  501     3.0     NaN        31.0     49.0  10301.0   
4      SI      1   21  501     3.0     NaN        31.0     49.0  10301.0   

  FireComp  ...  ZMCode  Sanborn   TaxMap  EDesigNum  APPBBL APPDate  \
0     L078  ...     NaN  501 017  50101.0        NaN     0.0     NaN   
1     L078  ...     NaN  501 017  50101.0        NaN     0.0     NaN   
2     L078  ...     NaN  501 017  50101.0        NaN     0.0     NaN   
3     L078  ...     NaN  501 017  50101.0        NaN     0.0     NaN   
4     L078  ...     NaN  501 017  50101.0        NaN     0.0     NaN   

  PLUTOMapID FIRM07_FLAG PFIRM15_FLAG Version  
0          1         NaN          NaN    18V1  
1          1         NaN          NaN    18V1  
2          1         NaN          NaN    18V1  
3          1         NaN          NaN    18V1  
4          1         NaN          NaN    18V1  

[5 rows x 87 columns]

# Quiz on Data Ingestion

In [9]:
# Question 1: The dataset contains complaints logged since what date?
df = df.sort_values("created_date", ascending=True)
df.head()

created_date  unique_key complaint_type  incident_zip  \
382008  2010-01-01T00:00:00.000    15636275       PLUMBING       10453.0   
381278  2010-01-01T00:00:00.000    15635168        HEATING       10453.0   
381279  2010-01-01T00:00:00.000    15635169        HEATING       10039.0   
381280  2010-01-01T00:00:00.000    15635170        HEATING       10467.0   
381281  2010-01-01T00:00:00.000    15635171        HEATING       10034.0   

                         incident_address                   street_name  \
382008                 54 EAST 179 STREET               EAST 179 STREET   
381278                  1740 GRAND AVENUE                  GRAND AVENUE   
381279  2866 FREDERICK DOUGLASS BOULEVARD  FREDERICK DOUGLASS BOULEVARD   
381280               3764 BRONX BOULEVARD               BRONX BOULEVARD   
381281                510 WEST 218 STREET               WEST 218 STREET   

       address_type      city  \
382008      ADDRESS     BRONX   
381278      ADDRESS     BRONX   
381279      ADDRESS  NEW YORK   
381280      ADDRESS     BRONX   
381281      ADDRESS  NEW YORK   

                                   resolution_description      borough  \
382008  The Department of Housing Preservation and Dev...  Unspecified   
381278  More than one complaint was received for this ...  Unspecified   
381279  The Department of Housing Preservation and Dev...  Unspecified   
381280  More than one complaint was received for this ...  Unspecified   
381281  More than one complaint was received for this ...  Unspecified   

         latitude  longitude              closed_date         location_type  \
382008  40.852125 -73.907442  2010-03-01T00:00:00.000  RESIDENTIAL BUILDING   
381278  40.849225 -73.913511  2010-01-03T00:00:00.000  RESIDENTIAL BUILDING   
381279  40.827575 -73.938443  2010-01-05T00:00:00.000  RESIDENTIAL BUILDING   
381280  40.883980 -73.866196  2010-01-02T00:00:00.000  RESIDENTIAL BUILDING   
381281  40.871461 -73.914744  2010-01-05T00:00:00.000  RESIDENTIAL BUILDING   

        status  
382008  Closed  
381278  Closed  
381279  Closed  
381280  Closed  
381281  Closed

In [10]:
# Question 2: The dataset contains complaints logged till what date?
df = df.sort_values("created_date", ascending=False)
df.head()

created_date  unique_key        complaint_type  incident_zip  \
1154  2020-07-02T23:58:10.000    46719688            WATER LEAK       10462.0   
354   2020-07-02T23:58:10.000    46715041         PAINT/PLASTER       10462.0   
321   2020-07-02T23:58:10.000    46715800            WATER LEAK       10462.0   
876   2020-07-02T23:58:10.000    46716681  UNSANITARY CONDITION       10462.0   
776   2020-07-02T23:51:38.000    46718294         PAINT/PLASTER       10473.0   

           incident_address       street_name address_type   city  \
1154  1450 PARKCHESTER ROAD  PARKCHESTER ROAD      ADDRESS  BRONX   
354   1450 PARKCHESTER ROAD  PARKCHESTER ROAD      ADDRESS  BRONX   
321   1450 PARKCHESTER ROAD  PARKCHESTER ROAD      ADDRESS  BRONX   
876   1450 PARKCHESTER ROAD  PARKCHESTER ROAD      ADDRESS  BRONX   
776       2045 STORY AVENUE      STORY AVENUE      ADDRESS  BRONX   

                                 resolution_description borough   latitude  \
1154  The following complaint conditions are still o...   BRONX  40.836319   
354   The following complaint conditions are still o...   BRONX  40.836319   
321   The following complaint conditions are still o...   BRONX  40.836319   
876   The following complaint conditions are still o...   BRONX  40.836319   
776   The following complaint conditions are still o...   BRONX  40.824831   

      longitude closed_date         location_type status  
1154 -73.859548         NaN  RESIDENTIAL BUILDING   Open  
354  -73.859548         NaN  RESIDENTIAL BUILDING   Open  
321  -73.859548         NaN  RESIDENTIAL BUILDING   Open  
876  -73.859548         NaN  RESIDENTIAL BUILDING   Open  
776  -73.854232         NaN  RESIDENTIAL BUILDING   Open

In [11]:
# Question 3: How many incidents have a missing Incident Address?
df[df["incident_address"].isnull()].shape #Check number of rows without address

(52831, 15)

In [12]:
# Question 4: How many valid ZIP Codes exist in the Bronx PLUTO dataset?
import numpy as np
df_BX_validzips = df_BX[df_BX["ZipCode"].notnull()]
len(np.unique(df_BX_validzips["ZipCode"]))

26

In [13]:
# Question 5: How many valid ZIP Codes exist in the Queens PLUTO dataset?
df_QN_validzips = df_QN[df_QN["ZipCode"].notnull()]
len(np.unique(df_QN_validzips["ZipCode"]))

65

# Question 1: What is the Top Complaint Type?

In [14]:
# To see the fields in df
print(df.shape)
print(df.columns)
df[0:1].transpose()

(6149696, 15)
Index(['created_date', 'unique_key', 'complaint_type', 'incident_zip',
       'incident_address', 'street_name', 'address_type', 'city',
       'resolution_description', 'borough', 'latitude', 'longitude',
       'closed_date', 'location_type', 'status'],
      dtype='object')


1154
created_date                                      2020-07-02T23:58:10.000
unique_key                                                       46719688
complaint_type                                                 WATER LEAK
incident_zip                                                        10462
incident_address                                    1450 PARKCHESTER ROAD
street_name                                              PARKCHESTER ROAD
address_type                                                      ADDRESS
city                                                                BRONX
resolution_description  The following complaint conditions are still o...
borough                                                             BRONX
latitude                                                          40.8363
longitude                                                        -73.8595
closed_date                                                           NaN
location_type                                        RESIDENTIAL BUILDING
status                                                               Open

In [15]:
# To see the fields in borough data
print(df_QN.shape)
print(df_QN.columns)
#df_QN[0:1].transpose()

(324583, 87)
Index(['Borough', 'Block', 'Lot', 'CD', 'CT2010', 'CB2010', 'SchoolDist',
       'Council', 'ZipCode', 'FireComp', 'PolicePrct', 'HealthCenterDistrict',
       'HealthArea', 'SanitBoro', 'SanitDistrict', 'SanitSub', 'Address',
       'ZoneDist1', 'ZoneDist2', 'ZoneDist3', 'ZoneDist4', 'Overlay1',
       'Overlay2', 'SPDist1', 'SPDist2', 'SPDist3', 'LtdHeight', 'SplitZone',
       'BldgClass', 'LandUse', 'Easements', 'OwnerType', 'OwnerName',
       'LotArea', 'BldgArea', 'ComArea', 'ResArea', 'OfficeArea', 'RetailArea',
       'GarageArea', 'StrgeArea', 'FactryArea', 'OtherArea', 'AreaSource',
       'NumBldgs', 'NumFloors', 'UnitsRes', 'UnitsTotal', 'LotFront',
       'LotDepth', 'BldgFront', 'BldgDepth', 'Ext', 'ProxCode', 'IrrLotCode',
       'LotType', 'BsmtCode', 'AssessLand', 'AssessTot', 'ExemptLand',
       'ExemptTot', 'YearBuilt', 'YearAlter1', 'YearAlter2', 'HistDist',
       'Landmark', 'BuiltFAR', 'ResidFAR', 'CommFAR', 'FacilFAR', 'BoroCode',
       'BBL', 'C

In [16]:
print(df.dtypes)

created_date               object
unique_key                  int64
complaint_type             object
incident_zip              float64
incident_address           object
street_name                object
address_type               object
city                       object
resolution_description     object
borough                    object
latitude                  float64
longitude                 float64
closed_date                object
location_type              object
status                     object
dtype: object


In [17]:
# Some Descriptive statistical analysis
df.describe(include=['object'])

created_date  complaint_type incident_address  \
count                   6149696         6149696          6096865   
unique                  1772055              30           185193   
top     2013-01-24T00:00:00.000  HEAT/HOT WATER  34 ARDEN STREET   
freq                       7581         1317699            14405   

            street_name address_type      city  \
count           6096865      6064944   6069403   
unique             6864            1        89   
top     GRAND CONCOURSE      ADDRESS  BROOKLYN   
freq              94464      6064944   2064552   

                                   resolution_description   borough  \
count                                             6141861   6149696   
unique                                                340         6   
top     The Department of Housing Preservation and Dev...  BROOKLYN   
freq                                              1731628   1777860   

                    closed_date         location_type   status  
count                   6016684               6096866  6149696  
unique                   805701                     5        5  
top     2012-11-07T00:00:00.000  RESIDENTIAL BUILDING   Closed  
freq                       7296               6089757  6009756

In [18]:
# Heat/Hot water and Heating seem to be the most frequent complaint types, they might even be related one to another.
df["complaint_type"].value_counts().head(10)

HEAT/HOT WATER          1317699
HEATING                  887869
PLUMBING                 722922
GENERAL CONSTRUCTION     500863
UNSANITARY CONDITION     469981
PAINT - PLASTER          361257
PAINT/PLASTER            355338
ELECTRIC                 312457
NONCONST                 260890
DOOR/WINDOW              212310
Name: complaint_type, dtype: int64

In [19]:
# Percentual incidence for top 5 complaint types
type_group = df.groupby("complaint_type")["unique_key"].count()/len(df) * 100
type_group.sort_values(ascending=False).head()

complaint_type
HEAT/HOT WATER          21.427059
HEATING                 14.437608
PLUMBING                11.755410
GENERAL CONSTRUCTION     8.144516
UNSANITARY CONDITION     7.642345
Name: unique_key, dtype: float64

Conclusion: HEAT/HOT WATER is the top complaint type, with 1,317,699 cases, accounting for 21,4% of total complaints from years 2010 to 2020.

# Quiz on complaint types

In [20]:
# Question 1: What is the total number of complaints that exist in the dataset?
len(df)

6149696

In [21]:
df1 = df[df["created_date"] < "2020-02-02"]

In [22]:
len(df1)

6019524

In [23]:
# Question 2: How many differnt Complaint Types can you find in the dataset, including duplicates entries of the same type?
len(np.unique(df["complaint_type"]))

30

In [26]:
# Question 3: How many Elevator complaints can you find in the dataset?
df1[df1["complaint_type"] =="ELEVATOR"]["complaint_type"].count()

6726

In [27]:
# Question 4: How many Electric complaints can you find in the dataset?
df1[df1["complaint_type"] =="ELECTRIC"]["complaint_type"].count()

307318

In [43]:
# Question 5: Using 800,000 as a threshold, what complaint type(s) 
#do you recommend the Department of Housing Preservation and Development of New York City address first? Select all that apply.
#df["complaint_type"].count()
df_group = pd.DataFrame(df["complaint_type"].value_counts())
df_group.columns = ["#complaints"]
df_group[df_group["#complaints"] >= 800000]

#complaints
HEAT/HOT WATER      1317699
HEATING              887869